test/test-parse-expression.cpp

```diff
@@ -3161,6 +3161,273 @@ TEST_F(test_parse_expression,
+  }
+
+  {
+    test_parser p(u8"['foo', g()]"_sv);
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_TRUE(side_effects);
+  }
+
+  {
+    test_parser p(u8"`foo`"_sv);
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_FALSE(side_effects);
+  }
+
+  {
+    test_parser p(u8"`foo ${bar}`"_sv);
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_FALSE(side_effects);
+  }
+
+  {
+    test_parser p(u8"`foo ${g()}`"_sv);
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_TRUE(side_effects);
+  }
+
+  {
+    test_parser p(u8"foo`bar`"_sv);
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_TRUE(side_effects);
+  }
+
+  {
+    test_parser p(u8"foo = bar"_sv);
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_TRUE(side_effects);
+  }
+
+  {
+    test_parser p(u8"x += y"_sv);
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_TRUE(side_effects);
+  }
+
+  {
+    test_parser p(u8"f() ? bar : 42"_sv);
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_TRUE(side_effects);
+  }
+
+  {
+    test_parser p(u8"foo ? f() : g()"_sv);
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_TRUE(side_effects);
+  }
+
+  {
+    test_parser p(u8"foo ? bar : g()"_sv);
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_TRUE(side_effects);
+  }
+
+  {
+    test_parser p(u8"foo ? f() : bar"_sv);
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_TRUE(side_effects);
+  }
+
+  {
+    test_parser p(u8"foo = bar ? x : y"_sv);
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_TRUE(side_effects);
+  }
+
+  {
+    test_parser p(u8"{foo: 42}"_sv);
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_FALSE(side_effects);
+  }
+
+  {
+    test_parser p(u8"{key: f()}"_sv);
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_TRUE(side_effects);
+  }
+
+  {
+    test_parser p(u8"{[f()]: true}"_sv);
+    expression* ast = p.parse_expression();
+    bool side_effects = p.parser().has_potential_side_effects(ast);
+    EXPECT_TRUE(side_effects);
```

Comment:
Nit: I think we can save on lines of code:
```suggestion
    EXPECT_TRUE(p.parser().has_potential_side_effects(ast));
```